src/Server.php

```diff
@@ -416,6 +420,80 @@ public function getPluginManager(){
 	}
 
+	/**
+	 * Provides an API implementation of $interface.
+	 *
+	 * $interface can be either an (abstract, open or final) class or interface.
+	 * $impl must implement $interface.
+	 *
+	 * The $interface is used to identify various API types,
+	 * and users should not try to provide APIs for two $interfaces that extend one another.
+	 * For example, PocketMine can provide a default BanList interface provides an interface to track user bans,
+	 * which calls `provideApi(BanList::class)`, but it does not call `provideApi(DefaultBanList::class)`.
+	 * Nevertheless, if alternative implementations are not intended,
+	 * it is fine to provide $interface as the final class.
+	 * (TODO: change "can provide" to "provides" when the BanList API is changed)
+	 *
+	 * If the declaration of `$interface` has a `@purpose` tag in its documentation,
+	 * it is provided as part of the error message to describe the purpose of the interface
+	 * when two conflicting implementations are provided,
+	 * e.g. `@purpose ban list` would result in the error:
+	 *
+	 * > Multiple plugins are providing ban list. Please disable one of them or check configuration
+	 *
+	 * The default implementation (usually provided by the module declaring the interface)
+	 * should call `provideDefaultApi` so that other plugins can override it without triggering errors.
+	 *
+	 * @template T of object
+	 * @param string $interface
+	 * @phpstan-param class-string<T> $interface
+	 * @param object $impl
+	 * @phpstan-param T $impl
+	 *
+	 * @throws \InvalidArgumentException if $impl is not an instance of $interface
+	 * @throws \RuntimeException if two non-default APIs are provided for the same interface
+	 *
+	 * @see #provideDefaultApi
+	 */
+	public function provideApi(string $interface, Plugin $plugin, object $impl) : void{
+		$this->apiMap->provideApi($interface, $plugin, $impl, false);
+	}
+
+	/**
+	 * Provides a *default* API implementation of $interface.
+	 *
+	 * `provideDefaultApi` must only be called exactly once, by the module that declared `$interface`.
+	 *
+	 * @template T of object
+	 * @param string $interface
+	 * @phpstan-param class-string<T> $interface
+	 * @param object $impl
+	 * @phpstan-param T $impl
+	 *
+	 * @throws \InvalidArgumentException if $impl is not an instance of $interface
+	 *
+	 * @see #provideApi for detailed semantics of API provision
+	 */
+	public function provideDefaultApi(string $interface, Plugin $plugin, object $impl) : void{
+		$this->apiMap->provideApi($interface, $plugin, $impl, true);
+	}
+
+	/**
+	 * Retrieves the current implementation of `getApi`.
+	 *
+	 * Callers can check whether this implementation is default by getting `$default` by reference.
+	 *
+	 * @template T of object
+	 * @param string $interface
+	 * @phpstan-param class-string<T> $interface
+	 * @param bool $default
+	 * @return object|null
+	 * @phpstan-return T|null
+	 */
+	public function getApi(string $interface, bool &$default = false) : ?object {
```

Comment:
Why is there an unsed parameter? _$default_ is never used

Reply:
Oops right, it's supposed to passthru to $default.

Reply:
Will you update or should I make a PR?

Reply:
just fyi, there is a "suggest change" button on GitHub, but I'm going to edit this myself for now

Reply:
Okay, fine.